### DISCUSSION TITLE
Provide code example for custom training.

### DISCUSSION THREAD

**Discussion by tahirahmad2030** [explosion/spaCy#6897] [Help: Coding & Implementations] [answered]
In previous versions of spacy, we had `examples` folder containing proper code on how to buld custom models on our own data for NER, Textcat etc. Please provide the same to V3.0.
## Your Environment

- Operating System: Ubuntu
- Python Version Used: 3.6
- spaCy Version Used: 3.0
- Environment Information:

**Comment by svlandeg** [comment]
Hi! I converted your issue to a discussion thread as we're trying to keep the issue tracker more focused on bug reports, while the discussions can be used for coding help, and general discussions.

**Comment by svlandeg** [comment, marked as answer]
If you want to upgrade your previous code examples, you can use the migration guide: https://spacy.io/usage/v3#migrating

In general though, we recommend using the [config system](https://spacy.io/usage/training#config) for training your custom components in v3, as this is much more powerful than the custom training scripts from v2. That is also why we removed the example scripts. 
Instead, you can now find some powerful examples in https://github.com/explosion/projects/tree/v3. More docs here: https://spacy.io/usage/projects

If you want to hit the ground running with your training in v3, I definitely recommend checking out the [`init config`](https://spacy.io/api/cli#init-config) command as well.

**Comment by honnibal** [comment]
To expand on this a little, the problem we constantly ran into from the `examples/` directory was that users would naturally go to it as the first place to look for how they should do something, e.g. train a model. But it's absolutely impossible to make a training script that's not either a) Really complicated, or b) Absolutely crap to use in practice.

Training takes a long time to run, so user-experience features like periodic accuracy reporting, progress, model checkpointing etc are pretty important. It's also important to be able to choose the GPU to use for some contexts. The batching and dropout also need to be specified, and the data needs to be loaded somehow, shuffled, and perhaps split into training and evaluation sections.

If you want to see how the training works in code, you can see the source for the [`spacy.training.loop.train`](https://github.com/explosion/spaCy/blob/master/spacy/training/loop.py) function. This is what the `spacy train` command calls into. We don't want to provide a separate, bad, implementation of model training in a separate script that we then tell people they shouldn't use.
